opacus/layers/dp_lstm.py

```diff
@@ -1,11 +1,55 @@
 import math
-from typing import List, Optional, Tuple
+from typing import List, Optional, Tuple, Union
 
 import torch
 import torch.nn as nn
+from torch.nn.utils.rnn import PackedSequence
 
 from .param_rename import ParamRenamedModule
 
+from torch.nn.utils.rnn import pad_sequence, pack_padded_sequence
+
+
+def compute_seq_lengths(batch_sizes: torch.Tensor):
+    max_batch_size = batch_sizes[0]
+    if len(batch_sizes) == 1:
+        return [1] * max_batch_size
+
+    running_seq = 0
+    running_seq_lengths = []
+    for i in range(1, len(batch_sizes)):
+        delta = batch_sizes[i - 1].item() - batch_sizes[i].item()
+        running_seq += 1
+        running_seq_lengths += delta * [running_seq]
+
+    running_seq += 1
+    running_seq_lengths += batch_sizes[-1].item() * [running_seq]
+    running_seq_lengths.reverse()
+    return running_seq_lengths
+
+
+def compute_last_states(h_n: torch.Tensor, c_n: torch.Tensor, seq_lengths: List):
+    max_batch_size = len(seq_lengths)
+    hidden_size = h_n[0].shape[-1]
+    h_last = torch.zeros(max_batch_size, hidden_size)
+    c_last = torch.zeros(max_batch_size, hidden_size)
+
+    for i, seq_len in enumerate(seq_lengths):
```

Comment:
This tensor creation via loop is very slow. A way to avoid it is to construct the indices first (you can use a loop for the indices), and then assign them all in one shot

Reply:
Actually, the type hints for ```h_n``` and ```c_n``` in ```compute_last_states``` are typos and they are both a list of tensors. I will fix that in my next commit. I think you suggested the above optimization thinking they are tensors. Being a list of tensors of varying shapes, I don't think we can apply the optimization you suggested. Is there a torch function for doing that? I couldn't find one.

Reply:
Yeah, not much to do then :) Should be fine